semistruct.go

```diff
@@ -0,0 +1,266 @@
+package semistruct
+
+import "strconv"
+import p "github.com/andyleap/parser"
+
+type Semistruct_log struct {
+	Priority int64
+	Tags     []string
+	Attrs    map[string]string
+}
+
+type kv_pair struct {
+	key string
+	val string
+}
+
+// Parse a semistructured log line into the semistruct_log struct
+// type.
+func ParseSemistruct() *p.Grammar {
+	o := p.And(
+		// Parse opening sentinel "!<"
+		OpenSentinel(),
+		SkipSpace(),
+
+		// Parse our log-level priority
+		p.Tag("priority", PriorityInt()),
+		SkipSpace(),
+
+		// Parse our tag list "[tag1:tag2:tag3]"
+		p.Tag("tags", Tags()),
+		SkipSpace(),
+
+		// Parse our attribute set "{ key=val key2=val2 }"
+		p.Tag("attrs", Attrs()),
+		SkipSpace(),
+
+		// Parse the ending sentinel ">!"
+		EndSentinel(),
+	)
+
+	o.Node(func(m p.Match) (p.Match, error) {
+		pr := p.GetTag(m, "priority").(int64)
```

Comment:
you need to use ok idiom here otherwise you may experience unexpected panics.

```
pr, ok := p.GetTag(m, "priority").(int64)
if !ok {
  // do something with error
}
```